app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java

```diff
@@ -0,0 +1,614 @@
+				roundedImageView.setOnClickListener(new View.OnClickListener() {
+					@Override
+					public void onClick(View v) {
+						TestFairy.showFeedbackForm();
+					}
+				});
+
+				container.addView(roundedImageView);
+			}
+		}
+
+		container.setPadding(padding, padding, padding, padding);
+		container.setLayoutParams(containerLayoutParams);
+		container.setOrientation(LinearLayout.VERTICAL);
+		container.setBackgroundColor(context.getResources().getColor(android.R.color.transparent));
+		container.setDividerPadding(0);
+		container.setShowDividers(LinearLayout.SHOW_DIVIDER_NONE);
+		root.addView(container);
+
+		if (visible) {
+			show(false);
+		} else {
+			hide(false);
+		}
+	}
+
+	private synchronized void removeOverlay(boolean postToContainer) {
+		Log.d(TAG, "Removing overlay");
+
+		if (container != null) {
+			Runnable remove = new Runnable() {
+				@Override
+				public void run() {
+					container.removeAllViews();
+					ViewParent parent = container.getParent();
+					if (parent instanceof ViewGroup) {
+						ViewGroup castedParent = (ViewGroup) parent;
+						castedParent.removeView(container);
+					}
+
+					container = null;
+				}
+			};
+
+			if (postToContainer) {
+				container.post(remove);
+			} else {
+				remove.run();
+			}
+		}
+
+		visible = false;
+	}
+
+	static private void onResume(Activity activity) {
+		Log.d(TAG, "TestFairyFeedbackOverlay is attaching to " + activity.getClass().getSimpleName());
+
+		if (instance != null && !(activity instanceof ProvideFeedbackActivity)) {
+			synchronized (instance) {
+				final View root = activity.getWindow().getDecorView().findViewById(android.R.id.content);
+				if (root != null && root instanceof FrameLayout) {
+					activity.runOnUiThread(new Runnable() {
+						@Override
+						public void run() {
+							instance.removeOverlay(false);
+							visible = visibleBeforePause;
+							instance.attachOverlay((FrameLayout) root);
+						}
+					});
+				} else {
+					activity.runOnUiThread(new Runnable() {
+						@Override
+						public void run() {
+							instance.removeOverlay(false);
+						}
+					});
+					Log.e(TAG, "There is no view to attach TestFairy feedback overlay");
+				}
+			}
+		} else if (!(activity instanceof ProvideFeedbackActivity)) {
+			instance = new TestFairyFeedbackOverlay();
+			final View root = activity.getWindow().getDecorView().findViewById(android.R.id.content);
+
+			if (root != null && root instanceof FrameLayout) {
+				activity.runOnUiThread(new Runnable() {
+					@Override
+					public void run() {
+						visible = visibleBeforePause;
+						instance.attachOverlay((FrameLayout) root);
+					}
+				});
+			} else {
+				Log.e(TAG, "There is no view to attach TestFairy feedback overlay");
+			}
+		}
+	}
+
+	static private void onPause(Activity activity) {
+		Log.d(TAG, "TestFairyFeedbackOverlay is detaching from " + activity.getClass().getSimpleName());
+
+		if (instance != null) {
```

Comment:
No need to nullify instance. Instance will be allocated if somebody clicks a button in the app to provide feedback. From here on now, it's in memory, it barely takes up any heap. It'll be easier to handle state machine if instance is never null.

Reply:
Allright 👍